pkg/controllers/provisioning/v1alpha1/reallocation/collect.go

```diff
@@ -0,0 +1,121 @@
+package reallocation
+
+import (
+	"context"
+	"fmt"
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha1"
+	f "github.com/awslabs/karpenter/pkg/utils/functional"
+	utilsnode "github.com/awslabs/karpenter/pkg/utils/node"
+	"github.com/awslabs/karpenter/pkg/utils/ptr"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+type Collector struct {
+	kubeClient client.Client
+}
+
+func (c *Collector) ManageUnderutilization(ctx context.Context, provisioner *v1alpha1.Provisioner) error {
+	// 1. Set TTL on TTLable Nodes
+	if err := c.markUnderutilized(ctx, provisioner); err != nil {
+		return fmt.Errorf("adding ttl and underutilized label, %w", err)
+	}
+
+	// 2. Remove TTL from Utilized Nodes
+	if err := c.clearUnderutilized(ctx, provisioner); err != nil {
+		return fmt.Errorf("removing ttl from node, %w", err)
+	}
+	return nil
+}
+
+// markUnderutilized adds a TTL to underutilized nodes
+func (c *Collector) markUnderutilized(ctx context.Context, provisioner *v1alpha1.Provisioner) error {
+	ttlable := []*v1.Node{}
+
+	// 1. Get all provisioner labeled nodes
+	nodes, err := c.getNodes(ctx, provisioner, map[string]string{})
+	if err != nil {
+		return err
+	}
+
+	// 2. Get nodes and pods for each node
+	for _, node := range nodes {
+		pods, err := c.getPodsOnNode(ctx, node.Name)
+		if err != nil {
+			zap.S().Debugf("Unable to get pods for node %s, %s", node.Name, err.Error())
+			continue
+		}
+		if utilsnode.IsUnderutilized(node, pods) {
+			if _, ok := node.Annotations[v1alpha1.ProvisionerTTLKey]; !ok {
+				ttlable = append(ttlable, node)
+			}
+		}
+	}
+
+	// 3. Set TTL for each underutilized node
+	for _, node := range ttlable {
+		persisted := node.DeepCopy()
+		node.Labels[v1alpha1.ProvisionerUnderutilizedKey] = "true"
+		node.Annotations[v1alpha1.ProvisionerTTLKey] = time.Now().Add(time.Duration(*provisioner.Spec.TTLSeconds) * time.Second).Format(time.RFC3339)
+		if err := c.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
+			zap.S().Debugf("Could not add TTL to node %s, %s", node.Name, err.Error())
```

Comment:
I think this is an error that we should bubble up for the same reasons as mentioned above.

Reply:
Agreed - see my other clarification